File: src/main/java/cub/hackerrank/tutorials/a30DaysOfCode/day18/day18.py
#!/usr/bin/env python3

import sys


# Write your code here
class Solution:
    def __init__(self):
        self.stack = [None for x in range(1024)]
        self.queue = [None for x in range(1024)]
        self.stackIdx = 0
        self.queueHead = 0
        self.queueTail = 0

    def pushCharacter(self, c):
        self.stack[self.stackIdx] = c
        self.stackIdx += 1

    def popCharacter(self):
        self.stackIdx -= 1
        return self.stack[self.stackIdx]

    def enqueueCharacter(self, c):
        self.queue[self.queueHead] = c
        self.queueHead += 1

    def dequeueCharacter(self):
        result = self.queue[self.queueTail]
        self.queueTail += 1
        return result


# read the string s
s = input()
# Create the Solution class object
obj = Solution()

l = len(s)
# push/enqueue all the characters of string s to stack
for i in range(l):
    obj.pushCharacter(s[i])
    obj.enqueueCharacter(s[i])

isPalindrome = True
'''
pop the top character from stack
dequeue the first character from queue
compare both the characters
'''
for i in range(l // 2):
    if obj.popCharacter() != obj.dequeueCharacter():
        isPalindrome = False
        break
# finally print whether string s is palindrome or not.
if isPalindrome:
    print("The word, " + s + ", is a palindrome.")
else:
    print("The word, " + s + ", is not a palindrome.")

File: src/main/java/cub/hackerrank/tutorials/a30DaysOfCode/day18/day18.js
/* eslint-disable */
process.stdin.resume()
process.stdin.setEncoding('ascii')

var input_stdin = ''
var input_stdin_array = ''
var input_currentline = 0

process.stdin.on('data', function (data) {
  input_stdin += data
})

process.stdin.on('end', function () {
  input_stdin_array = input_stdin.split('\n')
  main()
})

function readLine() {
  return input_stdin_array[input_currentline++]
}
/* eslint-enable */

// Write your code here
function Solution() {
  this.stack = []
  this.queue = []
  this.stackIdx = 0
  this.queueHead = 0
  this.queueTail = 0
  this.pushCharacter = function (c) {
    // eslint-disable-next-line
    this.stack[this.stackIdx++] = c
  }
  this.popCharacter = function () {
    // eslint-disable-next-line
    return this.stack[--this.stackIdx]
  }
  this.enqueueCharacter = function (c) {
    // eslint-disable-next-line
    this.queue[this.queueHead++] = c
  }
  this.dequeueCharacter = function () {
    // eslint-disable-next-line
    return this.queue[this.queueTail++]
  }
}

/* eslint-disable */
function main() {
  // read the string s
  const s = readLine()
  const len = s.length
  // create the Solution class object p
  const obj = new Solution()
  // push/enqueue all the characters of string s to stack
  for (var i = 0; i < len; i++) {
    obj.pushCharacter(s.charAt(i))
    obj.enqueueCharacter(s.charAt(i))
  }

  let isPalindrome = true
  /*
  pop the top character from stack
  dequeue the first character from queue
  compare both the characters */

  for (var i = 0; i < len / 2; i++) {
    if (obj.popCharacter() != obj.dequeueCharacter()) {
      isPalindrome = false
      break
    }
  }
  // finally print whether string s is palindrome or not

  if (isPalindrome) console.log(`The word, ${s}, is a palindrome.`)
  else console.log(`The word, ${s}, is not a palindrome.`)
}
/* eslint-enable */
